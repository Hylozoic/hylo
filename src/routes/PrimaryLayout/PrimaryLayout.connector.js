import { get, some } from 'lodash/fp'
import { connect } from 'react-redux'
import { matchPath } from 'react-router'
import { mobileRedirect } from 'util/mobile'
import getReturnToURL from 'store/selectors/getReturnToURL'
import resetReturnToURL from 'store/actions/resetReturnToURL'
import fetchForCurrentUser from 'store/actions/fetchForCurrentUser'
import fetchForGroup from 'store/actions/fetchForGroup'
import updateUserSettings from 'store/actions/updateUserSettings'
import { FETCH_FOR_CURRENT_USER, FETCH_FOR_GROUP } from 'store/constants'
import getMe from 'store/selectors/getMe'
import getGroupForCurrentRoute from 'store/selectors/getGroupForCurrentRoute'
import getMyMemberships from 'store/selectors/getMyMemberships'
import getSignupState, { SignupState } from 'store/selectors/getSignupState'
import isGroupRoute, { getSlugFromLocation } from 'store/selectors/isGroupRoute'
import { toggleDrawer, toggleGroupMenu } from './PrimaryLayout.store'
import getLastViewedGroup from 'store/selectors/getLastViewedGroup'

export function mapStateToProps (state, props) {
  const memberships = getMyMemberships(state, props)
  const showMenuBadge = some(m => m.newPostCount > 0, memberships)
  const hasMemberships = memberships.length > 0
  const slug = getSlugFromLocation(null, props)
  const group = getGroupForCurrentRoute(state, props)
  const lastViewedGroup = getLastViewedGroup(state)
  const currentGroupMembership = group && hasMemberships && memberships.find(m => m.group.id === group.id)
  const signupState = getSignupState(state)
  const signupInProgress = signupState === SignupState.InProgress

  let routeParams = { context: 'all' }
  const match = matchPath(props.location.pathname, { path: '/:context(groups)/:groupSlug/:view(events|groups|map|members|projects|settings|stream|topics)?' }) ||
                matchPath(props.location.pathname, { path: '/:context(all|public)/:view(events|groups|map|members|projects|settings|stream|topics)?' })
  if (match) {
    routeParams = match.params
  }

  return {
    currentUser: getMe(state),
    currentUserPending: state.pending[FETCH_FOR_CURRENT_USER],
    currentGroupMembership,
    downloadAppUrl: mobileRedirect(),
    group,
    groupPending: state.pending[FETCH_FOR_GROUP],
    hasMemberships,
    isDrawerOpen: get('PrimaryLayout.isDrawerOpen', state),
    isGroupMenuOpen: get('PrimaryLayout.isGroupMenuOpen', state),
    isGroupRoute: isGroupRoute(state, props),
    lastViewedGroup,
    returnToURL: getReturnToURL(state),
    routeParams,
    showMenuBadge,
    signupInProgress,
    slug
  }
}

export const mapDispatchToProps = {
  fetchForCurrentUser,
  fetchForGroup,
  resetReturnToURL,
  toggleDrawer,
  toggleGroupMenu,
  updateUserSettings
}

export default connect(mapStateToProps, mapDispatchToProps)
