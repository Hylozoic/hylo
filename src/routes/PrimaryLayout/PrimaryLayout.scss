.container {
  display: flex;
  flex-direction: column;
  align-content: stretch;
  align-items: stretch;
  height: 100vh;
  overflow: hidden;
  &.map-view {
    height: 100vh;
  }
}

.debug {
  .main {
    border: 1px dashed red;
  }
  .left {
    border: 1px dashed blue;
  }
  .center {
    border: 1px dashed green;
  }
  .sidebar {
    border: 1px dashed magenta;
  }
  .detail {
    border: 1px dashed purple;
  }
}

.top {
  /* empty */
  margin-bottom: -1 * $space-6x;
}

.main {
  padding-top: 56px;
  position: relative;
  flex-grow: 1;
  display: flex;
  flex-wrap: nowrap;
  width: 100%;
  max-width: 1500px;
  margin-left: auto;
  margin-right: auto;
  &.map-view {
    width: 100%;
    max-width: 100%;
  }
}

$drawer-width: 367px;
$drawer-height: 100vh;
$drawer-z-index: 51;
.drawer {
  width: $drawer-width;
  height: $drawer-height;
  position: fixed;
  left: 0px;
  z-index: 600;
  display: flex;
  flex-direction: column;
  transition-delay: 0s;
  transition-duration: 0.2s;
  box-shadow: 0px 4px 56px 0px rgba(35, 64, 91, .5);

  &.hidden {
    left: -$drawer-width;
    box-shadow: 0px 4px 56px 0px rgba(35, 64, 91, 0);
  }
}

.left {
  transition: all .45s ease;
}

.left .map-field {
    height: calc(100vh - 60px);
}

.center {
  padding-left: 3px;
  background: $color-white;
  // From Bootstrap make-col-ready():
  position: relative;
  min-height: 1px; // Prevent collapsing
  height: calc(100vh - 56px);
  flex: 0 0 $center-column-width;
  max-width: $center-column-width;
  overflow-y: auto;
  overflow-x: hidden;
  transition: all .45s ease;
  padding-right: $space-2x;
  padding-bottom: 100px;
  &.map-view {
    max-width: $map-center-column-width;
    flex: 0 0 $map-center-column-width;
    height: calc(100vh - 40px);
    padding-bottom: 0px;
  }
}

.singleColumn .center {
  position: absolute;
  left: 50%;
  transform: translateX(-50%);
  width: 100%;
  flex: 0 0 100%;
  max-width: 100%;
  padding-right: 0;
  height: calc(100vh - 58px);
  padding-left: 0;
}

.center.collapsedState {
  width: 43%;
  flex: 0 0 43%;
}

.center.map-view.collapsedState {
  width: calc(100% - 68px);
  max-width: calc(100% - 68px);
  flex: 0 0 100%;
}

.center::-webkit-scrollbar {
  background-color:transparent;
  width: 5px;
  border-radius: 5px;
}

.center::-webkit-scrollbar-thumb {
  background-color: rgba(200,200,200,.5   );
  outline: none;
  border-radius: 3px;
}

.sidebar {
  // From Bootstrap make-col-ready():
  position: relative;
  z-index: 10;
  min-height: 1px; // Prevent collapsing
  flex: 0 0 29%;
  max-width: 29%;
  transition-delay: 0.1s;
  transition-duration: 0.3s;
  top: 0;
  right: 0px;
  padding: 0 16px 0 $layout-spacing;
  overflow-y: auto;
  overflow-x: hidden;

  &.hidden {
    opacity: 0;
    right: -60px;
    transition-delay: 0.0s;
  }
}

.detail {
  background-color: white;
  position: absolute;
  z-index: 30;
  top: 58px;
  right: 0px;
  height: calc(100vh - 58px);
  width: $detail-column-width;
  box-shadow: 0px 4px 56px 0px $shadow-color;
  transition-delay: 0.1s;
  transition-duration: 0.2s;
  padding-bottom: 50px;
  overflow-y: auto;
  margin-top: -8px;

  &.hidden {
    opacity: 0;
    right: -1000px;
    transition-delay: 0.0s;
  }
}

.map-view .detail {
  max-width: 700px;
}

.detail::-webkit-scrollbar {
  background-color:transparent;
  width: 5px;
  border-radius: 5px;
}

.detail::-webkit-scrollbar-thumb {
  background-color: rgba(200,200,200,.5);
  outline: none;
}


@media screen and (max-width: 1024px) {
  .sidebar {
    top: -100vh;
    opacity: 0;
    transform: scale(0);
    height: 0;
  }

  .center {
    max-width: 70%;
    flex: 0 0 75%;
  }

  .center.collapsedState {
    width: 60%;
    flex: 0 0 60%;
  }

  .detail {
    width: 60%;
  }
}

@media screen and (max-width: 930px) {
  .center {
    max-width: 80%;
    flex: 0 0 80%;
    left: 5%;
  }

  .center.map-view {
    max-width: 93%;
    flex-basis: 93%;
    left: 0;
  }
}

@media screen and (max-width: 700px) {
  .center {
    max-width: 86%;
    flex: 0 0 86%;
    left: 0%;
  }

  .center.map-view {
    max-width: 90%;
    flex: 0 0 90%;
  }

  .detail {
    width: calc(100% - 80px);
    margin-top: 0;
  }

}

@media screen and (max-width: 600px) {
  .center {
    max-width: 100%;
    flex: 0 0 100%;
    left: 0%;
    height: calc(100vh - 104px);
    padding: 0 0 150px 0;
  }

  .center.map-view {
    max-width: 100%;
    flex: 0 0 100%;
    left: 0;
    padding: 0;
  }

  .detail {
    z-index: 110;
  }

  .left {
    position: fixed;
    box-shadow: 0px 4px 56px 0px $shadow-color;
    left: 0;
    bottom: 0;
    z-index: 100;
    width: 100%;
    max-width: 100%;
    flex: 0 0 100%;
    height: 50px;
  }
}

@media screen and (max-width: 425px) {
  .detail {
    width: 100%;
    margin-top: 0;
    top: 0;
  }

  .center {
    overflow-y: scroll;
  }
  .center.map-view {
    overflow-y: auto;
  }
}

@media screen and (max-width: 375px) {
  .drawer {
    width: 100%;

    &.hidden {
      left: -100%;
    }
  }
}
