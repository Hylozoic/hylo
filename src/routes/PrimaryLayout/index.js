import { withResizeDetector } from 'react-resize-detector'
import component from './PrimaryLayout'
import connector from './PrimaryLayout.connector'
export default connector(withResizeDetector(component, { handleHeight: false }))
