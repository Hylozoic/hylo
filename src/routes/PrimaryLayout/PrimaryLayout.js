import cx from 'classnames'
import { get, some } from 'lodash/fp'
import qs from 'querystring'
import React, { Component } from 'react'
import Intercom from 'react-intercom'
import Joyride from 'react-joyride'
import {
  matchPath,
  Redirect,
  Route,
  Switch
} from 'react-router-dom'

import config, { isTest } from 'config'
import Div100vh from 'react-div-100vh'
import AddLocation from 'routes/Signup/AddLocation'
import AllTopics from 'routes/AllTopics'
import CreateModal from 'components/CreateModal'
import GroupDetail from 'routes/GroupDetail'
import GroupSettings from 'routes/GroupSettings'
import GroupSidebar from 'routes/GroupSidebar'
import GroupWelcomeModal from 'routes/GroupWelcomeModal'
import Groups from 'routes/Groups'
import Drawer from './components/Drawer'
import Feed from 'routes/Feed'
import Stream from 'routes/Stream'
import MapExplorer from 'routes/MapExplorer'
import LandingPage from 'routes/LandingPage'
import Loading from 'components/Loading'
import MemberProfile from 'routes/MemberProfile'
// import MemberSidebar from 'routes/MemberSidebar'
import Members from 'routes/Members'
import Messages from 'routes/Messages'
import Navigation from './components/Navigation'
import NotFound from 'components/NotFound'
import PostDetail from 'routes/PostDetail'
import PostEditorModal from 'components/PostEditorModal'
import Welcome from 'routes/Signup/Welcome'
import Search from 'routes/Search'
import SignupModal from 'routes/Signup/SignupModal'
import SocketListener from 'components/SocketListener'
import SocketSubscriber from 'components/SocketSubscriber'
import TopNav from './components/TopNav'
import UploadPhoto from 'routes/Signup/UploadPhoto'
import UserSettings from 'routes/UserSettings'
import {
  OPTIONAL_POST_MATCH, OPTIONAL_GROUP_MATCH,
  OPTIONAL_NEW_POST_MATCH, POST_DETAIL_MATCH, GROUP_DETAIL_MATCH,
  REQUIRED_EDIT_POST_MATCH,
  isAboutPath,
  isSignupPath,
  isMapViewPath
} from 'util/navigation'
import { CENTER_COLUMN_ID, DETAIL_COLUMN_ID } from 'util/scrolling'
import './PrimaryLayout.scss'

// In order of more specific to less specific
const routesWithDrawer = [
  { path: `/:context(all|public)/:view(events|map|projects|stream)/${OPTIONAL_POST_MATCH}` },
  { path: `/:context(all|public)/:view(map)/${OPTIONAL_GROUP_MATCH}` },
  { path: `/:context(all|public)/${OPTIONAL_POST_MATCH}` },
  { path: '/:context(all)/:view(topics)/:topicName' },
  { path: '/:context(all)/:view(topics)' },
  // {/* Group Routes */}
  { path: `/:context(groups)/:groupSlug/:view(members)/:personId/${OPTIONAL_POST_MATCH}` },
  { path: `/:context(groups)/:groupSlug/:view(topics)/:topicName/${OPTIONAL_POST_MATCH}` },
  { path: `/:context(groups)/:groupSlug/:view(map)/${OPTIONAL_GROUP_MATCH}` },
  { path: `/:context(groups)/:groupSlug/:view(events|groups|map|members|projects|settings|stream|topics)/${OPTIONAL_POST_MATCH}` },
  { path: `/:context(groups)/:groupSlug/${OPTIONAL_POST_MATCH}` },
  // {/* Member Routes */}
  { path: `/:view(members)/:personId/${OPTIONAL_POST_MATCH}` },
  // {/* Other Routes */}
  { path: '/messages' },
  { path: '/settings' },
  { path: '/search' },
  { path: '/confirm-group-delete' }
]

const detailRoutes = [
  { path: `/:context(all|public)/:view(events|map|projects|stream)/${POST_DETAIL_MATCH}`, component: PostDetail },
  { path: `/:context(all|public)/:view(map)/${GROUP_DETAIL_MATCH}`, component: GroupDetail },
  { path: `/:context(all)/:view(members)/:personId/${POST_DETAIL_MATCH}`, component: PostDetail },
  { path: `/:context(all|public)/${POST_DETAIL_MATCH}`, component: PostDetail },
  { path: `/:context(groups)/:groupSlug/:view(events|map|projects|stream)/${POST_DETAIL_MATCH}`, component: PostDetail },
  { path: `/:context(groups)/:groupSlug/:view(members)/:personId/${POST_DETAIL_MATCH}`, component: PostDetail },
  { path: `/:context(groups)/:groupSlug/:view(map|groups)/${GROUP_DETAIL_MATCH}`, component: GroupDetail },
  { path: `/:context(groups)/:groupSlug/:view(topics)/:topicName/${POST_DETAIL_MATCH}`, component: PostDetail },
  { path: `/:context(groups)/:groupSlug/${POST_DETAIL_MATCH}`, component: PostDetail },
  { path: `/:view(members)/:personId/${POST_DETAIL_MATCH}`, component: PostDetail }
]

const createRoutes = [
  { path: `/:context(all|public)/:view(events|groups|map|projects|stream)/${OPTIONAL_POST_MATCH}` },
  { path: `/:context(all|public)/:view(members)/:personId/${OPTIONAL_POST_MATCH}` },
  { path: `/:context(all|public)/:views(topics)/:topicName/${OPTIONAL_POST_MATCH}` },
  { path: `/:context(all|public)/${OPTIONAL_POST_MATCH}` },
  { path: `/:context(groups)/:groupSlug/:view(members)/:personId/${OPTIONAL_POST_MATCH}` },
  { path: `/:context(groups)/:groupSlug/:view(events|groups|map|projects|stream|topics)/${OPTIONAL_POST_MATCH}` },
  { path: `/:context(groups)/:groupSlug/:view(topics)/:topicName/${OPTIONAL_POST_MATCH}` },
  { path: `/:context(groups)/:groupSlug/${OPTIONAL_POST_MATCH}` },
  { path: `/:view(members)/:personId/${OPTIONAL_POST_MATCH}` }
]

const signupRoutes = [
  { path: '/signup/upload-photo', child: UploadPhoto },
  { path: '/signup/add-location', child: AddLocation },
  { path: '/signup/welcome', child: Welcome }
]

const redirectRoutes = [
  // Redirects from old routes
  { from: '/:context(public|all)/p/:postId', to: '/:context/post/:postId' },
  { from: '/:context(public|all)/project', to: '/:context/projects' },
  { from: '/:context(public|all)/event', to: '/:context/events' },
  { from: '/p/:postId', to: '/all/post/:postId' },
  { from: '/m/:personId', to: '/all/members/:personId' },
  { from: '/m/:personId/p/:postId', to: '/all/members/:personId/post/:postId' },
  { from: '/all/m/:personId', to: '/all/members/:personId' },
  { from: '/all/m/:personId/p/:postId', to: '/all/members/:personId/post/:postId' },
  { from: '/(c|n)/:groupSlug/join/:accessCode', to: '/groups/:groupSlug/join/:accessCode' },
  { from: '/(c|n)/:groupSlug/', to: '/groups/:groupSlug/' },
  { from: '/(c|n)/:groupSlug/event', to: '/groups/:groupSlug/events' },
  { from: '/(c|n)/:groupSlug/event/:postId', to: '/groups/:groupSlug/events/post/:postId' },
  { from: '/(c|n)/:groupSlug/project', to: '/groups/:groupSlug/projects' },
  { from: '/(c|n)/:groupSlug/project/:postId', to: '/groups/:groupSlug/projects/post/:postId' },
  { from: '/(c|n)/:groupSlug/:view(members|map|settings|topics)', to: '/groups/:groupSlug/:view' },
  { from: '/(c|n)/:groupSlug/map/p/:postId', to: '/groups/:groupSlug/map/post/:postId' },
  { from: '/(c|n)/:groupSlug/p/:postId', to: '/groups/:groupSlug/post/:postId' },
  { from: '/(c|n)/:groupSlug/m/:personId', to: '/groups/:groupSlug/members/:personId' },
  { from: '/(c|n)/:groupSlug/m/:personId/p/:postId', to: '/groups/:groupSlug/members/:personId/post/:postId' },
  { from: '/(c|n)/:groupSlug/:topicName', to: '/groups/:groupSlug/topics/:topicName' },
  { from: '/(c|n)/:groupSlug/:topicName/p/:postId', to: '/groups/:groupSlug/topics/:topicName/post/:postId' },

  // redirects for context switching into global contexts, since these pages don't exist yet
  { from: '/all/(members|settings)', to: '/all' },
  { from: '/public/(members|topics|settings)', to: '/public' }
]

export default class PrimaryLayout extends Component {
  constructor (props) {
    super(props)
    this.state = {
      run: false,
      steps: [
        {
          disableBeacon: true,
          target: '#currentContext',
          title: 'You are here!',
          content: 'This is where we show you which group or view you are looking at. Hylo allows you to easily switch between groups as well as see updates from all your groups at once.'
        },
        {
          target: '#toggleDrawer',
          title: 'Switching groups & viewing all',
          content: 'By clicking on the group icon, you\'ll be able to switch between groups, or see all your groups at once.\n\nWant to see what else is out there? Navigate over to Public Groups & Posts to see!'
        },
        {
          target: '#groupMenu',
          title: 'Create & navigate',
          content: 'Here you can switch between types of content and create new content for people in your group or everyone on Hylo!'
        },
        {
          target: '#personalSettings',
          title: 'Messages, notifications & profile',
          content: 'Search for posts & people. Send messages to group members or people you see on Hylo. Stay up to date with current events and edit your profile.'
        }
      ],
      closeTheTour: false
    }
  }

  componentDidMount () {
    this.props.fetchForCurrentUser()
    if (this.props.slug) {
      this.props.fetchForGroup()
    }
  }

  componentDidUpdate (prevProps) {
    if (this.props.slug && this.props.slug !== prevProps.slug) {
      this.props.fetchForGroup()
    }
    if (this.props.location.pathname !== prevProps.location.pathname) {
      this.scrollToTop()
    }
  }

  handleClickStartTour = (e) => {
    e.preventDefault()
    this.props.updateUserSettings({ settings: { alreadySeenTour: true } })
    this.setState({
      run: true,
      closeTheTour: true
    })
  }

  closeTour = () => {
    this.props.updateUserSettings({ settings: { alreadySeenTour: true } })
    this.setState({
      closeTheTour: true
    })
  }

  closeDrawer = () => this.props.isDrawerOpen && this.props.toggleDrawer()

  scrollToTop = () => {
    const centerColumn = document.getElementById(CENTER_COLUMN_ID)
    if (centerColumn) {
      centerColumn.scrollTop = 0
    }
  }

  render () {
    const {
      currentGroupMembership,
      currentUser,
      group,
      groupPending,
      isDrawerOpen,
      isGroupRoute,
      location,
      showLogoBadge
    } = this.props

    if (!currentUser) {
      return <div styleName='container'>
        <Loading type='loading-fullscreen' />
      </div>
    }

    if (isGroupRoute) {
      if (!group && !groupPending) return <NotFound />
    }

    const queryParams = qs.parse(location.search.substring(1))
    const signupInProgress = get('settings.signupInProgress', currentUser)
    const hasDetail = some(
      ({ path }) => matchPath(location.pathname, { path }),
      detailRoutes
    )
    const isMapView = isMapViewPath(location.pathname)
    const collapsedState = hasDetail || (isMapView && queryParams['hideDrawer'] !== 'true')
    const isSingleColumn = (group && !currentGroupMembership) || matchPath(location.pathname, { path: '/members/:personId' })
    const showTourPrompt = !signupInProgress && !get('settings.alreadySeenTour', currentUser) && !isSingleColumn

    return <Div100vh styleName={cx('container', { 'map-view': isMapView, 'singleColumn': isSingleColumn, 'detailOpen': hasDetail })}>
      { showTourPrompt ? <Route path='/:context(all|public|groups)' component={props =>
        <div styleName={cx('tourWrapper', { 'tourClosed': this.state.closeTheTour })}>
          <div styleName='tourPrompt'>
            <div styleName='tourGuide'><img src='/axolotl-tourguide.png' /></div>
            <div styleName='tourExplanation'>
              <p><strong>Welcome to Hylo {currentUser.name}!</strong> I’d love to show you how things work, would you like a quick tour?</p>
              <p>To follow the tour look for the pulsing beacons! <span styleName='beaconExample'><span styleName='beaconA' /><span styleName='beaconB' /></span></p>
              <div>
                <button styleName='skipTour' onClick={this.closeTour}>No thanks</button>
                <button styleName='startTour' onClick={this.handleClickStartTour}>Show me Hylo</button>
              </div>
              <div styleName='speechIndicator' />
            </div>
          </div>
          <div styleName='tourBg' onClick={this.closeTour} />
        </div>} />
        : ' '}

      {/* Context navigation drawer */}
      <Switch>
        {routesWithDrawer.map(({ path }) => (
          <Route path={path} key={path} render={props => (
            <Drawer {...props} styleName={cx('drawer', { hidden: !isDrawerOpen })} {...{ group }} />
          )} />
        ))}
      </Switch>

      <TopNav styleName='top' onClick={this.closeDrawer} {...{ group, currentUser, showLogoBadge }} />

      <div styleName={cx('main', { 'map-view': isMapView })} onClick={this.closeDrawer}>
        {/* View navigation menu */}
        <Route path='/:context(all|public)' component={props =>
          <Navigation {...props}
            collapsed={collapsedState}
            styleName={cx('left', { 'map-view': isMapView })}
            mapView={isMapView}
          />}
        />
        {group && currentGroupMembership &&
          <Route path='/:context(groups)/:groupSlug' component={props =>
            <Navigation {...props}
              group={group}
              collapsed={collapsedState}
              styleName={cx('left', { 'map-view': isMapView })}
              mapView={isMapView}
            />}
          />
        }

        <Div100vh styleName={cx('center', { 'map-view': isMapView }, { collapsedState })} id={CENTER_COLUMN_ID}>
          <Switch>
            {redirectRoutes.map(({ from, to }) => <Redirect from={from} to={to} exact key={from} />)}
            {signupRoutes.map(({ path, child }) =>
              <Route path={path} key={path} render={props =>
                <SignupModal {...props} child={child} />} />)}
            {signupInProgress &&
              <RedirectToSignupFlow pathname={this.props.location.pathname} currentUser={currentUser} />}
            {!signupInProgress &&
              <RedirectToGroup path='/(|app)' currentUser={currentUser} />}
            {/* Member Routes */}
            <Route path={`/:view(members)/:personId/${OPTIONAL_POST_MATCH}`} render={props => <MemberProfile {...props} isSingleColumn={isSingleColumn} />} />
            <Route path={`/:context(all)/:view(members)/:personId/${OPTIONAL_POST_MATCH}`} component={MemberProfile} />
            {/* All and Public Routes */}
            <Route path={`/:context(all|public)/:view(stream)/${OPTIONAL_POST_MATCH}`} component={Stream} />
            <Route path={`/:context(all|public)/:view(events|projects)/${OPTIONAL_POST_MATCH}`} component={Feed} />
            <Route path={`/:context(all|public)/:view(map)/${OPTIONAL_POST_MATCH}`} component={MapExplorer} />
            <Route path={`/:context(all|public)/:view(map)/${OPTIONAL_GROUP_MATCH}`} component={MapExplorer} />
            <Route path='/:context(all|public)/:view(topics)/:topicName' component={Feed} />
            <Route path='/:context(all)/:view(topics)' component={AllTopics} />
            <Route path={`/:context(all|public)/${OPTIONAL_POST_MATCH}`} component={Feed} />
            {/* Group Routes */}
            {group && !currentGroupMembership &&
              <Route path={`/:context(groups)/:groupSlug`} render={props => <GroupDetail {...props} group={group} />} />}
            {currentGroupMembership && currentGroupMembership.settings.showJoinForm &&
              <Route path={`/:context(groups)/:groupSlug`} render={props => <GroupWelcomeModal {...props} group={group} />} />}
            <Route path={`/:context(groups)/:groupSlug/:view(map)/${OPTIONAL_POST_MATCH}`} component={MapExplorer} />
            <Route path={`/:context(groups)/:groupSlug/:view(map)/${OPTIONAL_GROUP_MATCH}`} component={MapExplorer} />
            <Route path={`/:context(groups)/:groupSlug/:view(stream)/${OPTIONAL_POST_MATCH}`} component={Stream} />
            <Route path={`/:context(groups)/:groupSlug/:view(events|projects)/${OPTIONAL_POST_MATCH}`} component={Feed} />
            <Route path='/:context(groups)/:groupSlug/:view(groups)' component={Groups} />
            <Route path={`/:context(groups)/:groupSlug/:view(members)/:personId/${OPTIONAL_POST_MATCH}`} component={MemberProfile} />
            <Route path='/:context(groups)/:groupSlug/:view(members)' component={Members} />
            <Route path={`/:context(groups)/:groupSlug/:view(topics)/:topicName/${OPTIONAL_POST_MATCH}`} component={Feed} />
            <Route path='/:context(groups)/:groupSlug/:view(topics)' component={AllTopics} />
            <Route path='/:context(groups)/:groupSlug/:view(settings)' component={GroupSettings} />
            <Route path={`/:context(groups)/:groupSlug/${OPTIONAL_POST_MATCH}`} exact component={LandingPage} />
            <Route path={`/:context(groups)/:groupSlug/${OPTIONAL_POST_MATCH}`} component={Feed} />
            {/* Other Routes */}
            <Route path='/settings' component={UserSettings} />
            <Route path='/search' component={Search} />
          </Switch>
        </Div100vh>
        {group && currentGroupMembership &&
          <div styleName={cx('sidebar', { hidden: (hasDetail || isMapView) })}>
            <Switch>
              <Route path={`/:context(groups)/:groupSlug/:view(events|map|groups|projects|stream)/${OPTIONAL_NEW_POST_MATCH}`} component={GroupSidebar} />
              <Route path={`/:context(groups)/:groupSlug/:view(topics)/:topicName/${OPTIONAL_NEW_POST_MATCH}`} component={GroupSidebar} />
              <Route path={`/:context(groups)/:groupSlug/${OPTIONAL_NEW_POST_MATCH}`} component={GroupSidebar} />
            </Switch>
          </div>
        }
        <div styleName={cx('detail', { hidden: !hasDetail })} id={DETAIL_COLUMN_ID}>
          <Switch>
            {detailRoutes.map(({ path, component }) => {
              return isAboutPath(location.pathname)
                ? <Route path={path} render={props => <GroupDetail {...props} communityId={group.id} />} key={path} />
                : <Route path={path} component={component} key={path} />
            })}
          </Switch>
        </div>
      </div>
      <Route path='/messages/:messageThreadId?' render={props => <Messages {...props} />} />
      <Switch>
        {createRoutes.map(({ path }) =>
          <Route path={path + '/create'} key={path + 'create'} children={({ match, location }) =>
            <CreateModal match={match} location={location} />} />)}
        {createRoutes.map(({ path }) =>
          <Route path={path + '/' + REQUIRED_EDIT_POST_MATCH} key={path + 'editpost'} children={({ match, location }) =>
            <PostEditorModal match={match} location={location} />} />)}
      </Switch>
      <SocketListener location={location} />
      <SocketSubscriber type='group' id={get('slug', group)} />
      <Intercom appID={isTest ? null : config.intercom.appId} hide_default_launcher />
      <Joyride
        run={this.state.run}
        continuous
        showProgress
        showClose
        tooltipComponent={TourTooltip}
        steps={this.state.steps}
      />
    </Div100vh>
  }
}

export function RedirectToSignupFlow ({ pathname }) {
  if (isSignupPath(pathname)) return null

  return <Redirect to='/signup/upload-photo' />
}

export function RedirectToGroup ({ path, currentUser }) {
  let redirectToPath = '/all'

  if (currentUser.memberships.count() > 0) {
    const mostRecentGroup = currentUser.memberships
      .orderBy(m => new Date(m.lastViewedAt), 'desc')
      .first()
      .group
    redirectToPath = `/groups/${mostRecentGroup.slug}`
  }

  return <Redirect exact from={path} to={redirectToPath} />
}

function TourTooltip ({
  continuous,
  index,
  step,
  backProps,
  closeProps,
  primaryProps,
  tooltipProps
}) {
  return <div {...tooltipProps} styleName='tooltipWrapper'>
    <div styleName='tooltipContent'>
      <div styleName='tourGuide'><img src='/axolotl-tourguide.png' /></div>
      <div>
        {step.title && <div styleName='stepTitle'>{step.title}</div>}
        <div>{step.content}</div>
      </div>
    </div>
    <div styleName='tooltipControls'>
      {index > 0 && (
        <button styleName='backButton' {...backProps}>
          Back
        </button>
      )}
      {continuous && (
        <button styleName='nextButton' {...primaryProps}>
          Next
        </button>
      )}
      {!continuous && (
        <button {...closeProps}>
          Close
        </button>
      )}
    </div>
  </div>
}
