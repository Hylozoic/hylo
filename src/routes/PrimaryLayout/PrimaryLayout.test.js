import { MemoryRouter, Redirect } from 'react-router'
import React from 'react'
import { shallow } from 'enzyme'
import orm from 'store/models'
import PrimaryLayout, {
  RedirectToCommunity
} from './PrimaryLayout'

it('shows NotFound if a currentUser is loaded and the community does not exist', () => {
  const wrapper = shallow(<PrimaryLayout
    isCommunityRoute
    currentUser={{}}
    location={{ pathname: '' }} />, { disableLifecycleMethods: true })
  expect(wrapper).toMatchSnapshot()
})

it('shows nothing for a community route if the community and currentUser are not loaded', () => {
  const wrapper = shallow(<PrimaryLayout
    isCommunityRoute
    communityPending />, { disableLifecycleMethods: true })
  expect(wrapper).toMatchSnapshot()
})

it('shows normal children for a community route if the community is loaded', () => {
  const wrapper = shallow(<PrimaryLayout
    isCommunityRoute
    location={{ pathname: '/', search: '' }}
    community={{ id: '1' }}
    currentUser={{}} />, { disableLifecycleMethods: true })
  expect(wrapper.name()).toEqual('div')
})

describe('RedirectToCommunity', () => {
  let session

  beforeEach(() => {
    session = orm.mutableSession(orm.getEmptyState())
    session.Me.create({ id: '1' })
  })

  it('sets `to` prop of Redirect correctly if currentUser has no memberships', () => {
    const currentUser = session.Me.first()
    const path = '/'
    const wrapper = shallow(<MemoryRouter>
      {RedirectToCommunity({ path, currentUser })}
    </MemoryRouter>)
    const expected = '/all'
    const actual = wrapper.find(Redirect).props().to
    expect(actual).toBe(expected)
  })

  it('sets `to` prop of Redirect correctly', () => {
    session.Community.create({ id: '1', slug: 'foo' })
    session.Membership.create({ id: '1', community: '1' })
    const currentUser = session.Me.first()
    currentUser.set('memberships', ['1'])
    const path = '/'
    const wrapper = shallow(<MemoryRouter>
      {RedirectToCommunity({ path, currentUser })}
    </MemoryRouter>)
    const actual = wrapper.find(Redirect).props().to
    const expected = '/c/foo'
    expect(actual).toBe(expected)
  })
})
