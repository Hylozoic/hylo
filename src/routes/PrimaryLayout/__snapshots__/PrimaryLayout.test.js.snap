// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shows NotFound if a currentUser is loaded and the group does not exist 1`] = `<withRouter(Connect(NotFound)) />`;

exports[`shows nothing for a group route if the group and currentUser are not loaded 1`] = `
<div
  data-stylename="container"
>
  <Loading
    type="loading-fullscreen"
  />
</div>
`;
